Improvements to ConstGlobalCache tag lists

kidder opened the pull request:
## Proposed changes

- Unify name used for tag lists
- Eliminate empty lists
- Tags from actions in phase-dependent action lists of components are collected in the metafunction that collects all ConstGlobalCache tags instead of in each component.

### Types of changes:

- [ ] Bugfix
- [x] New feature

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

kidder committed:
Update tests to not use OptionTags in ConstGlobalCache

kidder committed:
Simplify type aliases for ConstGlobalCache tags

This commit makes it easier to collect tags for objects in the
ConstGlobalCache from metavariables, components, and actions.  If no
tags are added by a metavariable, component, or action, the empty type list
can be omitted.  Also the collection of tags from actions in the
phase-dependent action list is done in helper metafunctions rather than
explictly done in each component.  The following changes were made:
- Rename const_global_cache_tag_list to const_global_cache_tags so there is
a consistent type alias in metavariables, components, and actions.
- const_global_cache_tags with empty lists are removed.

In addition, some clean up was done in ParallelComponentHelpers.

nilsdeppe commented on src/Parallel/ConstGlobalCache.hpp:
```diff
@@ -69,9 +69,33 @@ using get_component_if_mocked = tmpl::front<tmpl::type_from<tmpl::conditional_t<
+///   `allocate_array` function of an array component.  The type alias may be
+///   omitted for an empty list.
+/// - `Action::const_global_cache_tags` for each `Action` in the
+///    `phase_dependent_action_list` of each `Component` of
```
Unsure if losing my mind, but is the indentation changing here? 4 spaces vs. 3?

nilsdeppe commented on src/Parallel/ConstGlobalCache.hpp:
```diff
@@ -69,9 +69,33 @@ using get_component_if_mocked = tmpl::front<tmpl::type_from<tmpl::conditional_t<
+/// The tag list for the items added to the ConstGlobalCache is created by
+/// combining the following tag lists:
+/// - `Metavariables::const_global_cache_tags` which should contain only those
+///    tags that cannot be added from the following tag lists:
```
having the colon at the end makes it seem like the next two items should be sub-elements. Is this intended or should it just say something like `from the lists below.`?

nilsdeppe commented on src/Parallel/ParallelComponentHelpers.hpp:
```diff
@@ -20,66 +20,69 @@ struct get_inbox_tags_from_action<Action,
-struct get_const_global_cache_tags_from_action {
+namespace detail {
+template <class ParallelStruct, class = cpp17::void_t<>>
+struct get_const_global_cache_tags_from_parallel_struct {
```
What *is* a `parallel_struct`? A C++ comment would be helpful :)

nilsdeppe commented on src/Parallel/ParallelComponentHelpers.hpp:
```diff
@@ -20,66 +20,69 @@ struct get_inbox_tags_from_action<Action,
+            Metavariables>::type,
+        tmpl::transform<
+            typename Metavariables::component_list,
+            tmpl::bind<detail::component_const_global_cache_tags, tmpl::_1>>,
```
Isn't this the same as `get_const_global_cache_tags_from_parallel_struct<tmpl::_1>`?

nilsdeppe commented on src/Parallel/ParallelComponentHelpers.hpp:
```diff
@@ -20,66 +20,69 @@ struct get_inbox_tags_from_action<Action,
-    typename Parallel_detail::get_const_global_cache_tags_from_action<
-        Action>::type;
+template <typename Component>
+using component_const_global_cache_tags =
```
Since all these eager metafunctions are turned into lazy metafunctions below, why not just make them lazy right away?

nilsdeppe commented on docs/DevGuide/Parallelization.md:
```diff
@@ -194,12 +195,12 @@ Each %Parallel Component struct must have the following type aliases:
-   `Parallel::get_const_global_cache_tags` metafunction. These tags
-   correspond to items that are stored in the ConstGlobalCaches (of
-   which there is one copy per Charm++ node).
+5. `using const_global_cache_tags` is set to a `tmpl::list` of tags
```
What if I want `allocate_array` to use something but I don't want it in the cache?

nilsdeppe commented on docs/DevGuide/CreatingExecutables.md:
```diff
@@ -83,13 +83,14 @@ the `PrintMessage` action is called during the `Execute` phase.
+Executables can read in an input file (specified by the `--input-file`
+argument) that will be parsed when the executable begins.  %Options
+specified in the input file can be used to either place items in a
+constant global cache (by specifying tags in the
```
`constant global cache` -> `ConstGlobalCache`?

kidder committed:
Improve documentation

Add clarifications about const_global_cache_tags.
Add namespaces so DataBox and ConstGlobalCache are linked to their
documentation.

kidder commented:
@nilsdeppe fixup pushed

nilsdeppe reviewed (commented):
LGTM, happy for you to squash :+1:

kidder commented:
@markscheel please look at

markscheel commented on docs/DevGuide/CreatingExecutables.md:
```diff
@@ -85,8 +85,8 @@ component is created upon, and prints a message.
-specified in the input file can be used to either place items in a
-constant global cache (by specifying tags in the
+specified in the input file can be used to either place items in the
+Parallel::ConstGlobalCache (by specifying tags in the
```
It sounds like there are only two ways that options specified in the input file can get into the executable. But there's a 3rd way: specifying initialization_tags.

markscheel commented on docs/DevGuide/Parallelization.md:
```diff
@@ -172,18 +172,19 @@ Each %Parallel Component struct must have the following type aliases:
+   last. Simple actions (described below) can be executed in any
+   phase. If there are no iterable actions in a phase then a
+   `PhaseAction` need not be specified for that phase. However, at
+   least one `PhaseAction`, even it if is empty, must be specified.
```
typo: 'it if' -> 'if it'

markscheel commented on docs/DevGuide/Parallelization.md:
```diff
@@ -192,15 +193,18 @@ Each %Parallel Component struct must have the following type aliases:
+   Parallel::ConstGlobalCache (of which there is one copy per Charm++ node).
+   The alias can be omitted if the list is empty.  (See
+   `array_allocation_tags` below for specifying tags needed for the
+   `allocate_array` function that will not create items for the
```
'function that will not create items for the' -> 'function; these `array_allocation_tags` will not be put in the'

markscheel commented on docs/DevGuide/Parallelization.md:
```diff
@@ -293,27 +297,30 @@ and methods using template parameters and invocation of actions. This approach
+difference between the two calls is that by calling an action through
+the parallel component you will first finish the series of actions you
+are in, then when they are complete Charm++ will call the next queued
+action.
```
I realize that this was a whitespace change only, but perhaps this should be clarified if people other than me think this is unclear. I have a few questions (for my own understanding):

1. Is the only difference between the two calls that the bad one uses `Parallel::get_parallel_component<ParallelComponent>(cache).ckLocal()` and the good one (which is not spelled out explicitly) uses `Parallel::get_parallel_component<ParallelComponent>(cache)` without the `ckLocal()`?  
2. Is the call with the `ckLocal()` executed immediately? (instead of deferred and queued)
3. Is the problem that the action is calling itself, or is it equally bad to call ActionA from ActionB using `ckLocal()` if ActionA and ActionB are in the same ParallelComponent?
4. Under what circumstances can `ckLocal()` be used legitimately?  Presumably you cannot use it to call an Action on a different ParallelComponent.

kidder replied:
1) yes; the former directly calls the local object, the latter is handled by Charm++
2) yes
3) calling any action is disallowed (this is stated on line 308 and tested in Test_AlgorithmNestedApply2.cpp)

markscheel reviewed (changes requested):
A few documentation comments.

kidder commented:
@markscheel I pushed a fixup and answered your questions (let me know if you want changes in the doc based on the answers)

markscheel reviewed (dismissed):
Thanks!  Please squash.

I also have questions as to the difference between `ckLocal()` and `ckLocalBranch()` (and what are legitimate usages of each) but that is getting very very tangential to this PR.

kidder commented:
@markscheel @nilsdeppe squashed

nilsdeppe commented:
Travis GCC6 debug timed out while backing up cache in final build. All tests passed.
